frame/contracts/src/tests.rs

```diff
@@ -4383,3 +4383,114 @@ fn delegate_call_indeterministic_code() {
 		);
 	});
 }
+
+#[test]
+#[cfg(feature = "unstable-interface")]
+fn reentrant_count_works_with_call() {
+	let (wasm, code_hash) = compile_module::<Test>("reentrant_count_call").unwrap();
+	let contract_addr = Contracts::contract_address(&ALICE, &code_hash, &[]);
+
+	ExtBuilder::default().existential_deposit(100).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		assert_ok!(Contracts::instantiate_with_code(
+			RuntimeOrigin::signed(ALICE),
+			300_000,
+			GAS_LIMIT,
+			None,
+			wasm,
+			vec![],
+			vec![],
+		));
+
+		// passing reentrant count to the input
+		let input = 0.encode();
+
+		Contracts::bare_call(
+			ALICE,
+			contract_addr,
+			0,
+			GAS_LIMIT,
+			None,
+			input,
+			true,
+			Determinism::Deterministic,
+		)
+		.result
+		.unwrap();
+	});
+}
+
+#[test]
+#[cfg(feature = "unstable-interface")]
+fn reentrant_count_works_with_delegated_call() {
+	let (wasm, code_hash) = compile_module::<Test>("reentrant_count_delegated_call").unwrap();
+	let contract_addr = Contracts::contract_address(&ALICE, &code_hash, &[]);
+
+	ExtBuilder::default().existential_deposit(100).build().execute_with(|| {
+		let _ = Balances::deposit_creating(&ALICE, 1_000_000);
+
+		assert_ok!(Contracts::instantiate_with_code(
+			RuntimeOrigin::signed(ALICE),
+			300_000,
+			GAS_LIMIT,
+			None,
+			wasm,
+			vec![],
+			vec![],
+		));
+
+		// adding a callstack height to the input
+		let input = (code_hash, 1).encode();
+
+		Contracts::bare_call(
+			ALICE,
+			contract_addr.clone(),
+			0,
+			GAS_LIMIT,
+			None,
+			input,
+			true,
+			Determinism::Deterministic,
+		)
+		.result
+		.unwrap();
+	});
+}
+
+#[test]
+#[cfg(feature = "unstable-interface")]
+fn account_entrance_count_works() {
```

Comment:
This test is insufficient:

- It tests only for its own address: Please additionally test for the address of **another** contract on the call stack.
- This only test for the value `1`. Please also test at least for `0`, too.

Reply:
The test still tests the trivial case where nothing is on the call stack. But let's not be too strict here.